### DISCUSSION TITLE
Bug with explicit Mapping hints?

### DISCUSSION THREAD

**Discussion by dbushong** [microsoft/pyright#9554] [General]
```python
from typing import Mapping

x: Mapping[int, int | str] = {1: 2}

reveal_type(x)

x[3] = "four"
```

In this example code, no error is produced on line 3, but on line 5 the type of `x` is given as `dict[int, int]`.  On line 7 this is underscored when you finally get an error:

```
Argument of type "Literal['four']" cannot be assigned to parameter "value" of type "int" in function "__setitem__"
  "Literal['four']" is not assignable to "int"  (reportArgumentType)
```

See: https://pyright-play.net/?pythonVersion=3.14&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCyAhgsugFDkAeAXISWWgNqowA0mK%2BAPlAM4wQAXSgBeKAG8AjHQBMAX0ogApgDdlRADYB9eAmUAKKgEpKVJgGYR4gETAwAVxA2gA

It seems to me that either:

1. The `: Mapping[...]` type hint on line 3 should work, and the type of `x` should be `Mapping[int, int | str]` and no assignments should be allowed **or**
2. An error should be emitted on line 3 saying that it isn't allowed

**Comment by erictraut** [comment]
This is related to a [recent discussion](https://discuss.python.org/t/interoperability-issue-with-module-level-inline-annotations/72992/12) on the typing forum. That discussion involved a bunch of (unrelated) issues, but it raised the same question you're raising above.

The question is whether bidirectional type inference for the expression `{1: 2}` should result in `dict[int, int]` or `dict[int, int | str]`. Mypy produces the latter whereas pyright produces the former. 

I need to give this some additional thought, but I'm not convinced that pyright's behavior is correct in this case.

**Comment by dbushong** [comment]
Iiiiinteresting.  Thanks!  I'll follow the discussion with interest.
